audio/pwaudio.c

```diff
@@ -0,0 +1,787 @@
+    }
+    default:
+        break;
+    }
+}
+
+//add notification for when the stream parameter changes
+static void on_stream_param_changed(void *_data, uint32_t id, const struct spa_pod *param)
+{
+	pwaudio *pw = _data;
+    const struct spa_pod *params[1];
+    uint8_t buffer[1024];
+    struct spa_pod_builder b = SPA_POD_BUILDER_INIT(buffer, sizeof(buffer));
+
+	/* NULL means to clear the format */
+	if (param == NULL || id != SPA_PARAM_Format)
+		return;
+
+	if (spa_format_parse(param, &pw->format.media_type, &pw->format.media_subtype) < 0)
+		return;
+
+	/* only accept raw audio */
+	if (pw->format.media_type != SPA_MEDIA_TYPE_audio ||
+	    pw->format.media_subtype != SPA_MEDIA_SUBTYPE_raw)
+		return;
+
+	/* call a helper function to parse the format for us. */
+	spa_format_audio_raw_parse(param, &pw->format.info.raw);
+
+    params[0] = spa_pod_builder_add_object(&b,
+                    SPA_TYPE_OBJECT_ParamBuffers, SPA_PARAM_Buffers,
+                    SPA_PARAM_BUFFERS_blocks,     SPA_POD_Int(1),
+                    SPA_PARAM_BUFFERS_size,       SPA_POD_CHOICE_RANGE_Int(
+							        BUFFER_SAMPLES * sizeof(float), 32, INT32_MAX),
+			        SPA_PARAM_BUFFERS_stride,  SPA_POD_Int(sizeof(float)));
+    if (!params[0]) {
+        pw_log_error("Could not build parameter pod\n");
+        return;
+    }
+
+    if (pw_stream_update_params(pw->stream, params, 1) < 0) {
+        pw_log_error("Could not update stream parameters\n");
+        return;
+    }
+}
+
+static const struct pw_stream_events capture_stream_events = {
+	PW_VERSION_STREAM_EVENTS,
+	.destroy = capture_stream_destroy,
+    .state_changed = on_stream_state_changed,
+    .param_changed = on_stream_param_changed,
+	.process = capture_on_process
+};
+
+static const struct pw_stream_events playback_stream_events = {
+    PW_VERSION_STREAM_EVENTS,
+    .destroy = playback_stream_destroy,
+    .state_changed = on_stream_state_changed,
+    .param_changed = on_stream_param_changed,
+    .process = playback_on_process
+};
+
+static size_t qpw_read(HWVoiceIn *hw, void *data, size_t len)
+{
+    PWVoiceIn *pw = (PWVoiceIn *) hw;
+    pwaudio *c = pw->g;
+    const char *error = NULL;
+    size_t l;
+
+    pw_thread_loop_lock(c->thread_loop);
+    if (pw_stream_get_state (pw->stream, &error) != PW_STREAM_STATE_STREAMING) {
+        /* wait for stream to become ready */
+        l = 0;
+        goto done_unlock;
+    }
+    capture_on_process(&pw->stream);
+
+done_unlock:
+    pw_thread_loop_unlock (c->thread_loop);
+    return l;
+}
+
+static size_t qpw_write(HWVoiceOut *hw, void *data, size_t len)
+{
+    PWVoiceOut *pw = (PWVoiceOut *) hw;
+    pwaudio *c = pw->g;
+    const char *error = NULL;
+    size_t l;
+
+    pw_thread_loop_lock(c->thread_loop);
+    if (pw_stream_get_state (pw->stream, &error) != PW_STREAM_STATE_STREAMING) {
+        /* wait for stream to become ready */
+        l = 0;
+        goto done_unlock;
+    }
+    playback_on_process(&pw->stream);
```

Comment:
Same here, you need to write len beytes of data in the ringbuffer (that will then be read from by pipewire in the capture_on_process callback).